In [1]:
%matplotlib inline
import pyspark
import pandas as pd
import math
from pyspark.ml.feature import OneHotEncoder, StringIndexer, Normalizer
from pyspark.ml.pipeline import Transformer
from pyspark.ml.param.shared import HasInputCol, HasOutputCol, Param
from pyspark import keyword_only
from pyspark.ml import Pipeline
from pyspark.sql import Row
from pyspark.sql import SQLContext

In [2]:
sc = pyspark.SparkContext(master='local[*]')
sqlContext = SQLContext(sc)

In [3]:
pd.set_option('display.max_columns', 500)

In [4]:
shelter_pd = pd.read_csv('train.csv')

In [5]:
shelter_pd

AnimalID        Name             DateTime      OutcomeType  \
0      A671945     Hambone  2014-02-12 18:22:00  Return_to_owner   
1      A656520       Emily  2013-10-13 12:44:00       Euthanasia   
2      A686464      Pearce  2015-01-31 12:28:00         Adoption   
3      A683430         NaN  2014-07-11 19:09:00         Transfer   
4      A667013         NaN  2013-11-15 12:52:00         Transfer   
5      A677334        Elsa  2014-04-25 13:04:00         Transfer   
6      A699218       Jimmy  2015-03-28 13:11:00         Transfer   
7      A701489         NaN  2015-04-30 17:02:00         Transfer   
8      A671784        Lucy  2014-02-04 17:17:00         Adoption   
9      A677747         NaN  2014-05-03 07:48:00         Adoption   
10     A668402         NaN  2013-12-05 15:50:00         Transfer   
11     A666320         NaN  2013-11-04 14:48:00         Adoption   
12     A684601      Rocket  2016-02-03 11:27:00         Adoption   
13     A704702     Scooter  2015-06-08 16:30:00  Return_to_owner   
14     A688584     Preston  2015-11-25 15:00:00  Return_to_owner   
15     A678825      Oliver  2014-07-12 12:10:00         Adoption   
16     A678050         NaN  2014-05-03 16:15:00         Transfer   
17     A680028         NaN  2014-06-07 12:54:00         Transfer   
18     A679010     Chrissy  2014-05-17 11:32:00         Transfer   
19     A684787        Kuma  2014-07-30 17:34:00  Return_to_owner   
20     A670658      Anders  2014-01-19 15:03:00         Transfer   
21     A708812    Munchkin  2015-09-18 15:19:00         Transfer   
22     A705262     Jamaica  2015-08-15 14:22:00         Adoption   
23     A665781     Quattro  2013-10-28 16:32:00         Adoption   
24     A676333     Trigger  2014-04-09 17:44:00  Return_to_owner   
25     A712750       Olive  2015-10-03 15:44:00         Transfer   
26     A718832         Nia  2016-01-15 17:31:00         Adoption   
27     A698777       Chloe  2015-03-25 18:50:00         Adoption   
28     A715348     Cymbale  2015-11-21 13:01:00         Adoption   
29     A708666         Jax  2015-07-30 14:30:00  Return_to_owner   
...        ...         ...                  ...              ...   
26699  A676460       Mario  2014-04-21 14:01:00       Euthanasia   
26700  A704999        Lion  2015-06-15 19:28:00  Return_to_owner   
26701  A678842        Ella  2014-06-15 17:41:00         Transfer   
26702  A713637         NaN  2015-10-11 09:42:00             Died   
26703  A716670      Mowmow  2015-12-04 12:22:00         Transfer   
26704  A716223       Alice  2015-11-17 17:17:00         Transfer   
26705  A662498       Grace  2013-10-19 15:34:00         Transfer   
26706  A689942        Zoey  2014-10-19 13:29:00         Adoption   
26707  A682690         NaN  2014-07-01 17:06:00       Euthanasia   
26708  A666920      Herbie  2013-11-13 17:32:00         Transfer   
26709  A714386         NaN  2015-10-24 00:00:00         Adoption   
26710  A692622      Bonnie  2014-11-24 17:21:00         Transfer   
26711  A666284       Sonja  2013-10-30 18:32:00         Transfer   
26712  A535277      Merlin  2015-04-20 16:04:00  Return_to_owner   
26713  A671169         NaN  2014-01-20 17:37:00       Euthanasia   
26714  A679031        Pete  2014-05-31 16:11:00         Adoption   
26715  A708833        Luna  2015-08-05 17:03:00         Adoption   
26716  A697158         Max  2015-05-02 21:04:00         Adoption   
26717  A519949       Bella  2014-06-30 17:34:00  Return_to_owner   
26718  A701044         NaN  2015-04-28 14:26:00         Adoption   
26719  A707871         NaN  2015-07-20 09:00:00         Transfer   
26720  A704405     Brownie  2015-07-18 14:08:00         Adoption   
26721  A662019  Green Bean  2014-07-17 09:43:00         Adoption   
26722  A687056         NaN  2014-08-31 09:00:00         Transfer   
26723  A716997        Zeus  2016-01-29 18:52:00         Adoption   
26724  A702446         NaN  2015-05-14 11:56:00         Transfer   
26725  A718934         NaN  2016-01-20 18:59:00         Transf

In [6]:
shelter_pd.describe(include = "all")

AnimalID   Name             DateTime OutcomeType OutcomeSubtype  \
count     26729  19038                26729       26729          13117   
unique    26729   6374                22918           5             16   
top     A705677    Max  2015-08-11 00:00:00    Adoption        Partner   
freq          1    136                   19       10769           7816   

       AnimalType SexuponOutcome AgeuponOutcome                   Breed  \
count       26729          26728          26711                   26729   
unique          2              5             44                    1380   
top           Dog  Neutered Male         1 year  Domestic Shorthair Mix   
freq        15595           9779           3969                    8810   

              Color  
count         26729  
unique          366  
top     Black/White  
freq           2824

El Data Quality Report nos da un análisis de los datos. Tenemos 26729 filas en el dataset y se puede ver que hay un missing value en la columna SexuponOutcome (<1%), 18 en AgeuponOutcome (<1%), 7691 en Name (~28.7%) y 13612 en OutcomeType(~50.9%).

Como vamos a manejar el problema de los missing values?

- En el caso de SexuponOutcome, podemos simplemente borrar las filas porque no va a afectar nuestro modelo de predicciones.
- En el caso de AgeuponOutcome, vamos a borrar las filas también porque son muy pocos datos (<1%).
- En el caso de Name, no nos importa porque no vamos a usar esa columna en nuestro modelo. (El nombre no influye el Outcome del animal..)
- En el caso de OutcomeSubtype, tenemos muchos missing value porque por algunos valores de OutcomeType, como por ejemplo <i>Return_to_owner</i>, no hacía falta poner más información.

Que tipo de transformaciones vamos a aplicar sobre los datos para que se pueda usar para crear un modelo de predicción?
- Name: Borramos la columna porque no nos interesa en nuestras predicciones.
- DateTime: Borramos la hora y el día porque no es relevante aquí y ponemos un incremental que empieza con el primer mes del dataset.
- OutcomeType: Es un feature categórica entonces vamos a aplicar el StringIndexer para que lo transforme a feature numérica.
- OutcomeSubtype: Borramos la columna porque no nos interesa en nuestras predicciones.
- AnimalType: Es un feature categórica entonces vamos a aplicar el StringIndexer para que lo transforme a feature numérica.
- SexuponOutcome: Es un feature categórica entonces vamos a aplicar el StringIndexer para que lo transforme a feature numérica.
- AgeuponOutcome: Vamos a convertir cada edad en semana para que todos los datos tengan la misma MEDIDA??
- Breed: Es un feature categórica entonces vamos a aplicar el StringIndexer para que lo transforme a feature numérica.
- Color: Es un feature categórica entonces vamos a aplicar el StringIndexer para que lo transforme a feature numérica.

Ahora vamos a limpiar los datos y transformar cada feature categórica a numérica para que se pueda usar para crear un modelo de predicción.

### Part I: Cleaning the data

In [7]:
def valueCount(DF, column, value):## Function that counts the number of occurrences of a value in a given column
    count = 0
    columnIndex = DF.columns.get_loc(column)
    for (index, row) in DF.iterrows():
        if row[columnIndex] == value:
            count += 1
    return count

def missingValuesRemover(DF, column):## Removes the missing values in the given column
    count = 0
    columnIndex = DF.columns.get_loc(column)
    for (index, row) in DF.iterrows():
        if str(row[columnIndex]).lower().split() == ['nan']:
            DF = DF.drop(index)
            count += 1
    print(str(count) + " element(s) removed")
    return DF

NOTE:
- Should we normalise each feature in the range [0, 1]?

In [8]:
shelter_pd = missingValuesRemover(shelter_pd, "AgeuponOutcome")
shelter_pd = missingValuesRemover(shelter_pd, "SexuponOutcome")
del shelter_pd['AnimalID']
del shelter_pd['OutcomeSubtype']
del shelter_pd['Name']

18 element(s) removed
1 element(s) removed


### Part II: Indexation

In [9]:
shelter_pd[['DateTime', 'OutcomeType', 'AnimalType', 'SexuponOutcome', 'AgeuponOutcome', 'Breed', 'Color']] = shelter_pd[['DateTime', 'OutcomeType', 'AnimalType', 'SexuponOutcome', 'AgeuponOutcome', 'Breed', 'Color']].astype(str)
shelter_df = sqlContext.createDataFrame(shelter_pd)

## Creating a Spark Dataframe from the Pandas Dataframe in order to use it with the Decision Tree model.

In [10]:
from pyspark.sql.functions import UserDefinedFunction
from pyspark.sql.types import IntegerType, LongType, DoubleType, StringType


class AgeuponOutcomeTransformer(Transformer, HasInputCol, HasOutputCol): 
    # Does binning using the module of the division
    @keyword_only
    def __init__(self, inputCol=None, outputCol=None):
        super(AgeuponOutcomeTransformer, self).__init__()
        kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)
        
    @keyword_only 
    def setParams(self, inputCol=None, outputCol=None):
        kwargs = self.setParams._input_kwargs
        return self._set(**kwargs)
        
    def _transform(self, dataset):
        in_col = dataset[self.getInputCol()]
        out_col = self.getOutputCol()
        udf = UserDefinedFunction(lambda x: category_func(x), StringType())
        return dataset.withColumn(out_col, udf(in_col))

def category_func(in_value):
    if (in_value):
        age_in_days = 0
        splited_value = in_value.split(' ')        
        if (splited_value[1] == 'years' or splited_value[1] == 'year'):
            age_in_days = int(splited_value[0])*365
        elif(splited_value[1] == 'months' or splited_value[1] == 'month'):
            age_in_days = int(splited_value[0])*30
        elif(splited_value[1] == 'weeks' or splited_value[1] == 'week'):
            age_in_days = int(splited_value[0])*7
                
        if (age_in_days > 0 and age_in_days <= 365):
            return 'young'
        elif (age_in_days > 365 and age_in_days <= 365*3):
            return 'young_adult'
        elif (age_in_days > 365*3 and age_in_days <= 365*7):
            return 'adult'
        elif (age_in_days > 365*7 and age_in_days <= 365*10):
            return 'old_adult'
        elif (age_in_days > 365*10):
            return 'old'
    else:
        return 'None'

In [11]:
class BinningTransformer(Transformer, HasInputCol, HasOutputCol): 
    # Does binning using the module of the division
    @keyword_only
    def __init__(self, inputCol=None, outputCol=None):
        super(BinningTransformer, self).__init__()
        kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)
        
    @keyword_only 
    def setParams(self, inputCol=None, outputCol=None):
        kwargs = self.setParams._input_kwargs
        return self._set(**kwargs)
        
    def _transform(self, dataset):
        in_col = dataset[self.getInputCol()]
        out_col = self.getOutputCol()
        udf = UserDefinedFunction(lambda x: bin_func(x), DoubleType())
        return dataset.withColumn(out_col, udf(in_col))

def bin_func(in_value):
    if (in_value):
        age_in_days = 0
        splited_value = in_value.split(' ')        
        if (splited_value[1] == 'years' or splited_value[1] == 'year'):
            age_in_days = int(splited_value[0])*365
        elif(splited_value[1] == 'months' or splited_value[1] == 'month'):
            age_in_days = int(splited_value[0])*30
        elif(splited_value[1] == 'weeks' or splited_value[1] == 'week'):
            age_in_days = int(splited_value[0])*7
                
        if (age_in_days > 0 and age_in_days <= 365):
            return 1.0
        elif (age_in_days > 365 and age_in_days <= 365*3):
            return 2.0
        elif (age_in_days > 365*3 and age_in_days <= 365*7):
            return 3.0
        elif (age_in_days > 365*7 and age_in_days <= 365*10):
            return 4.0
        elif (age_in_days > 365*10):
            return 5.0
    else:
        return 0.0

In [12]:
## Here are all the indexers we're gonna use in the pipeline:
OutcomeType_indexer = StringIndexer(inputCol="OutcomeType", outputCol="OutcomeType_idx")
AnimalType_indexer = StringIndexer(inputCol="AnimalType", outputCol="AnimalType_idx")
SexuponOutcome_indexer = StringIndexer(inputCol="SexuponOutcome", outputCol="SexuponOutcome_idx")
Breed_indexer = StringIndexer(inputCol="Breed", outputCol="Breed_idx")
Color_indexer = StringIndexer(inputCol="Color", outputCol="Color_idx")
#Age_indexer = StringIndexer(inputCol='AgeuponOutcome', outputCol='AgeuponOutcome_idx')
Bin_transformer = BinningTransformer(inputCol='AgeuponOutcome', outputCol='AgeuponOutcome_binned')
AgeuponOutcome_transformer = AgeuponOutcomeTransformer(inputCol='AgeuponOutcome', outputCol='AgeuponOutcome_category')


## Should we use OneHotEncoder or StringIndexer for continuous features?
## It looks like we're using StringIndexer more for stuff like boolean values and
## OneHotEncoder for actual categorical features.. But not sure
## Anyway if it's actally OneHotEncoder that we need to use, we just need to add
## One OneHotIndexer and use the OutcomeType_idx we just created as inputCol..

In [13]:
pipeline = Pipeline(stages=[OutcomeType_indexer, AnimalType_indexer, SexuponOutcome_indexer, Breed_indexer, Color_indexer, AgeuponOutcome_transformer, Bin_transformer])
model = pipeline.fit(shelter_df)

In [14]:
shelter_df_idx = model.transform(shelter_df)
shelter_df_idx.show()

+-------------------+---------------+----------+--------------+--------------+--------------------+------------------+---------------+--------------+------------------+---------+---------+-----------------------+---------------------+
|           DateTime|    OutcomeType|AnimalType|SexuponOutcome|AgeuponOutcome|               Breed|             Color|OutcomeType_idx|AnimalType_idx|SexuponOutcome_idx|Breed_idx|Color_idx|AgeuponOutcome_category|AgeuponOutcome_binned|
+-------------------+---------------+----------+--------------+--------------+--------------------+------------------+---------------+--------------+------------------+---------+---------+-----------------------+---------------------+
|2014-02-12 18:22:00|Return_to_owner|       Dog| Neutered Male|        1 year|Shetland Sheepdog...|       Brown/White|            2.0|           0.0|               0.0|    100.0|      5.0|                  young|                  1.0|
|2013-10-13 12:44:00|     Euthanasia|       Cat| Spayed Fema

In [15]:
shelter_df_only_idx = shelter_df_idx
shelter_df_only_idx = shelter_df_only_idx.drop('OutcomeType', 'AnimalType', 'SexuponOutcome', 'Breed', 'Color')

## Dropping the tables that are still categorical because they're no use to us.

In [16]:
shelter_df_only_idx.registerTempTable("shelter")
sqlContext.sql("SELECT OutcomeType_idx, AnimalType_idx, SexuponOutcome_idx, Breed_idx, Color_idx, AgeuponOutcome_binned FROM shelter").show()

## Showing all the indexed tables

+---------------+--------------+------------------+---------+---------+---------------------+
|OutcomeType_idx|AnimalType_idx|SexuponOutcome_idx|Breed_idx|Color_idx|AgeuponOutcome_binned|
+---------------+--------------+------------------+---------+---------+---------------------+
|            2.0|           0.0|               0.0|    100.0|      5.0|                  1.0|
|            3.0|           1.0|               1.0|      0.0|     32.0|                  1.0|
|            0.0|           0.0|               0.0|      1.0|      9.0|                  2.0|
|            1.0|           1.0|               2.0|      0.0|    122.0|                  1.0|
|            1.0|           0.0|               0.0|    281.0|     13.0|                  2.0|
|            1.0|           0.0|               3.0|    152.0|     10.0|                  1.0|
|            1.0|           1.0|               2.0|      0.0|     21.0|                  1.0|
|            1.0|           1.0|               4.0|      0.0

#################################
#################################


OKAY GUYS, 

Tenemos que cambiar la columna fecha y edad para poder usarlo con decision tree


#################################
#################################

### Para realizar naive base convertimos el cleaned dataframe de spark a pandas

In [17]:
shelter_df_only_category = shelter_df_idx
shelter_df_only_category = shelter_df_only_category.drop('DateTime', 'OutcomeType_idx', 'AnimalType_idx', 'SexuponOutcome_idx', 'Breed_idx', 'Color_idx','AgeuponOutcome_binned', 'Ageuponoutcome')
shelter_cleaned_df = shelter_df_only_category.toPandas()
shelter_cleaned_df

OutcomeType AnimalType SexuponOutcome  \
0      Return_to_owner        Dog  Neutered Male   
1           Euthanasia        Cat  Spayed Female   
2             Adoption        Dog  Neutered Male   
3             Transfer        Cat    Intact Male   
4             Transfer        Dog  Neutered Male   
5             Transfer        Dog  Intact Female   
6             Transfer        Cat    Intact Male   
7             Transfer        Cat        Unknown   
8             Adoption        Dog  Spayed Female   
9             Adoption        Dog  Spayed Female   
10            Transfer        Cat        Unknown   
11            Adoption        Dog  Spayed Female   
12            Adoption        Dog  Neutered Male   
13     Return_to_owner        Dog  Neutered Male   
14     Return_to_owner        Dog  Neutered Male   
15            Adoption        Cat  Neutered Male   
16            Transfer        Cat    Intact Male   
17            Transfer        Dog  Intact Female   
18            Transfer        Cat  Intact Female   
19     Return_to_owner        Dog  Neutered Male   
20            Transfer        Dog    Intact Male   
21            Transfer        Cat  Spayed Female   
22            Adoption        Dog  Spayed Female   
23            Adoption        Dog  Neutered Male   
24     Return_to_owner        Dog  Neutered Male   
25            Transfer        Cat  Intact Female   
26            Adoption        Dog  Spayed Female   
27            Adoption        Dog  Spayed Female   
28            Adoption        Cat  Spayed Female   
29     Return_to_owner        Cat  Neutered Male   
...                ...        ...            ...   
26680       Euthanasia        Dog  Neutered Male   
26681  Return_to_owner        Dog    Intact Male   
26682         Transfer        Dog  Intact Female   
26683             Died        Cat  Spayed Female   
26684         Transfer        Cat  Neutered Male   
26685         Transfer        Cat  Intact Female   
26686         Transfer        Cat  Intact Female   
26687         Adoption        Cat  Spayed Female   
26688       Euthanasia        Cat        Unknown   
26689         Transfer        Dog  Neutered Male   
26690         Adoption        Dog  Spayed Female   
26691         Transfer        Dog  Intact Female   
26692         Transfer        Dog  Intact Female   
26693  Return_to_owner        Dog  Neutered Male   
26694       Euthanasia        Cat  Intact Female   
26695         Adoption        Cat  Spayed Female   
26696         Adoption        Dog  Spayed Female   
26697         Adoption        Dog  Neutered Male   
26698  Return_to_owner        Dog  Spayed Female   
26699         Adoption        Dog  Neutered Male   
26700         Transfer        Cat        Unknown   
26701         Adoption        Cat  Spayed Female   
26702         Adoption        Cat  Spayed Female   
26703         Transfer        Cat    Intact Male   
26704         Adoption        Dog  Neutered Male   
26705         Transfer        Cat    Intact Male   
26706         Transfer        Cat  Spayed Female   
26707         Adoption        Dog  Neutered Male   
26708         Transfer        Cat    Intact Male   
26709         Transfer        Cat    Intact Male   

                                       Breed                Color  \
0                      Shetland Sheepdog Mix          Brown/White   
1                     Domestic Shorthair Mix          Cream Tabby   
2                               Pit Bull Mix           Blue/White   
3                     Domestic Shorthair Mix           Blue Cream   
4                Lhasa Apso/Miniature Poodle                  Tan   
5          Cairn Terrier/Chihuahua Shorthair            Black/Tan   
6                     Domestic Shorthair Mix           Blue Tabby   
7                     Domestic Shorthair Mix          Brown Tabby   
8              American Pit Bull Terrier Mix            Red/White   
9                              Cairn Terrier                White   
10                    Domes